test/dotnet-add-proj.Tests/GivenDotnetAddProj.cs

```diff
@@ -0,0 +1,236 @@
+        [InlineData("Lib", "", false)]
+        [InlineData("LibWithProjectGuid", "{84A45D44-B677-492D-A6DA-B3A71135AB8E}", false)]
+        [InlineData("Lib", "", true)]
+        [InlineData("LibWithProjectGuid", "{84A45D44-B677-492D-A6DA-B3A71135AB8E}", true)]
+        public void ItAddsProjectAndPrintsStatus(string projectName, string projectGuid, bool makeRelative)
+        {
+            var setup = CreateTestSetup(TestGroup, ProjectName);
+            string slnDir = Path.Combine(setup.TestRoot, "MoreThanOne");
+            string slnName = "App.sln";
+
+            var projectToAdd = makeRelative
+                ? $"\"..{Path.DirectorySeparatorChar}{projectName}{Path.DirectorySeparatorChar}{projectName}.csproj\""
+                : $"\"{setup.GetProjectFullPath(projectName)}\"";
+
+            var cmd = new AddProjCommand()
+                .WithWorkingDirectory(slnDir)
+                .WithSolution(slnName)
+                .Execute(projectToAdd);
+            cmd.Should().Pass();
+            cmd.StdOut.Should().Contain("added to the solution");
+            cmd.StdErr.Should().BeEmpty();
+
+            var slnFile = new SlnFile();
+            slnFile.Read(Path.Combine(slnDir, slnName));
+            VerifyProjectInSolution(slnFile, projectName, projectGuid);
+
+            Restore(Path.Combine(setup.TestRoot, "App"), "App.csproj");
+            Build(slnDir, slnName);
+        }
+
+        [Fact]
+        public void WhenSolutionAlreadyContainsProjectItDoesntDuplicate()
+        {
+            var setup = CreateTestSetup(TestGroup, ProjectName);
+            string slnDir = Path.Combine(setup.TestRoot, "MoreThanOne");
+            string slnName = "AppAndLib.sln";
+            string slnFullPath = Path.Combine(slnDir, slnName);
+
+            var contentBefore = File.ReadAllText(slnFullPath);
+            var cmd = new AddProjCommand()
+                .WithWorkingDirectory(slnDir)
+                .WithSolution(slnName)
+                .Execute($"\"{setup.GetProjectFullPath("Lib")}\"");
+            cmd.Should().Pass();
+            cmd.StdOut.Should().Contain("already contains project");
+            cmd.StdErr.Should().BeEmpty();
+
+            var contentAfter = File.ReadAllText(slnFullPath);
+            contentAfter.Should().BeEquivalentTo(contentBefore);
+        }
+
+        [Fact]
+        public void WhenPassedMultipleProjectsAndOneOfthemDoesNotExistItCancelsWholeOperation()
+        {
+            var setup = CreateTestSetup(TestGroup, ProjectName);
+            string slnDir = Path.Combine(setup.TestRoot, "MoreThanOne");
+            string slnName = "App.sln";
+            string slnFullPath = Path.Combine(slnDir, slnName);
+
+            var contentBefore = File.ReadAllText(slnFullPath);
+            var cmd = new AddProjCommand()
+                .WithWorkingDirectory(slnDir)
+                .WithSolution(slnName)
+                .Execute($"\"{setup.GetProjectFullPath("Lib")}\" \"IDoNotExist.csproj\"");
```

Comment:
@blackdwarf thoughts here? Should we fail the full operation, or add the ones that exist and report on the ones we failed to add? While I like the determinism here, I can imagine being frustrated if I have to navigate through an already long string to fix something broken in the middle.

I also think we should talk about the "add many" UX pattern. It creates a bunch of ambiguous UX decisions like this one, and I cannot imagine myself using it except in scripts... which may as well invoke the command multiple times...